liualgotrader/trading/base.py

```diff
@@ -91,6 +93,43 @@ async def run(self) -> Optional[asyncio.Task]:
 
     async def close(self):
         ...
+      
+    def get_broker(self):
+        return config.broker
+        
+    def broker_working_limits(self, start_date, end_date):
+        market_opens, market_closes = get_market_schedule()
+        hours = market_closes - market_opens     
+        
+        days = len(get_trading_days(start_date, end_date).index)
+        return hours, days
+        
+    def max_data_points_per_load(self):
+        broker = get_broker()
+        if broker == BrokerType.alpaca:
+            max_data_points_per_load = 10000 
+        elif broker == BrokerType.gemini:
+            max_data_points_per_load = 500
+        return max_data_points_per_load
+    
+    def calculate_data_range(self,
+          start_date: date,
+          end_date: date, 
+          ) -> List[Optional[pd.DatetimeIndex]]:
+        
+        hours, days = broker_working_limits(start_date, end_date)
+
+        scale_factor_hours = hours
+        scale_factor_minutes = scale_factor_hours * 60
+
+        data_points = scale_factor_hours if self.scale == TimeScale.day else scale_factor_minutes
```

Comment:
@ksilo you should note the `codecov` warnings on your PR. It means that there are no tests, in the tests directory that cover these lines & flows. It's *very* important that you add unit-tests to changes you do, to validate the fixes automatically. We do not have a fancy QA team, that we can ask them to run tests and make sure nothing got broken. So we need to rely on pytest (= unit-tests). If you're not familiar with the concept, let's discuss it.